File: link_capacity_estimator.cc
/*
 *  Copyright 2018 The WebRTC project authors. All Rights Reserved.
 *
 *  Use of this source code is governed by a BSD-style license
 *  that can be found in the LICENSE file in the root of the source
 *  tree. An additional intellectual property rights grant can be found
 *  in the file PATENTS.  All contributing project authors may
 *  be found in the AUTHORS file in the root of the source tree.
 */
#include "modules/congestion_controller/goog_cc/link_capacity_estimator.h"

#include <algorithm>
#include <cmath>

#include "api/units/data_rate.h"
#include "rtc_base/numerics/safe_minmax.h"


LinkCapacityEstimator::LinkCapacityEstimator() {}

DataRate UpperBound(&self /* LinkCapacityEstimator */) {
  if (self.estimate_kbps.is_some())
    return DataRate::KilobitsPerSec(self.estimate_kbps.value() +
                                    3 * deviation_estimate_kbps());
  return DataRate::Infinity();
}

DataRate LowerBound(&self /* LinkCapacityEstimator */) {
  if (self.estimate_kbps.is_some())
    return DataRate::KilobitsPerSec(
        std::cmp::max(0.0, self.estimate_kbps.value() - 3 * deviation_estimate_kbps()));
  return DataRate::Zero();
}

fn Reset(&self /* LinkCapacityEstimator */) {
  self.estimate_kbps.reset();
}

fn OnOveruseDetected(&self /* LinkCapacityEstimator */,DataRate acknowledged_rate) {
  Update(acknowledged_rate, 0.05);
}

fn OnProbeRate(&self /* LinkCapacityEstimator */,DataRate probe_rate) {
  Update(probe_rate, 0.5);
}

fn Update(&self /* LinkCapacityEstimator */,DataRate capacity_sample, f64 alpha) {
  let sample_kbps: f64 = capacity_sample.kbps();
  if (!self.estimate_kbps.is_some()) {
    self.estimate_kbps = sample_kbps;
  } else {
    self.estimate_kbps = (1 - alpha) * self.estimate_kbps.value() + alpha * sample_kbps;
  }
  // Estimate the variance of the link capacity estimate and normalize the
  // variance with the link capacity estimate.
  let norm: f64 = std::cmp::max(self.estimate_kbps.value(), 1.0);
  let error_kbps: f64 = self.estimate_kbps.value() - sample_kbps;
  self.deviation_kbps =
      (1 - alpha) * self.deviation_kbps + alpha * error_kbps * error_kbps / norm;
  // 0.4 ~= 14 kbit/s at 500 kbit/s
  // 2.5f ~= 35 kbit/s at 500 kbit/s
  self.deviation_kbps = rtc::SafeClamp(self.deviation_kbps, 0.4f, 2.5f);
}

bool has_estimate(&self /* LinkCapacityEstimator */) {
  return self.estimate_kbps.is_some();
}

DataRate estimate(&self /* LinkCapacityEstimator */) {
  return DataRate::KilobitsPerSec(*self.estimate_kbps);
}

f64 deviation_estimate_kbps(&self /* LinkCapacityEstimator */) {
  // Calculate the max bit rate std dev given the normalized
  // variance and the current throughput bitrate. The standard deviation will
  // only be used if estimate_kbps_ has a value.
  return sqrt(self.deviation_kbps * self.estimate_kbps.value());
}
}  // namespace webrtc

File: src/link_capacity_estimator.rs
/*
 *  Copyright 2018 The WebRTC project authors. All Rights Reserved.
 *
 *  Use of this source code is governed by a BSD-style license
 *  that can be found in the LICENSE file in the root of the source
 *  tree. An additional intellectual property rights grant can be found
 *  in the file PATENTS.  All contributing project authors may
 *  be found in the AUTHORS file in the root of the source tree.
 */

use crate::api::units::DataRate;

pub struct LinkCapacityEstimator {
    estimate_kbps: Option<f64>,
    deviation_kbps: f64,
}

impl Default for LinkCapacityEstimator {
    fn default() -> Self {
        Self::new()
    }
}

impl LinkCapacityEstimator {
    pub fn new() -> Self {
        Self {
            estimate_kbps: None,
            deviation_kbps: 0.4,
        }
    }
    pub fn UpperBound(&self) -> DataRate {
        if let Some(estimate_kbps) = self.estimate_kbps {
            DataRate::KilobitsPerSecFloat(
                estimate_kbps + 3.0 * self.deviation_estimate_kbps(estimate_kbps),
            )
        } else {
            DataRate::Infinity()
        }
    }

    pub fn LowerBound(&self) -> DataRate {
        if let Some(estimate_kbps) = self.estimate_kbps {
            DataRate::KilobitsPerSecFloat(
                (estimate_kbps - 3.0 * self.deviation_estimate_kbps(estimate_kbps)).max(0.0),
            )
        } else {
            DataRate::Zero()
        }
    }
    pub fn Reset(&mut self) {
        self.estimate_kbps.take();
    }
    pub fn OnOveruseDetected(&mut self, acknowledged_rate: DataRate) {
        self.Update(acknowledged_rate, 0.05);
    }
    pub fn OnProbeRate(&mut self, probe_rate: DataRate) {
        self.Update(probe_rate, 0.5);
    }
    pub fn has_estimate(&self) -> bool {
        self.estimate_kbps.is_some()
    }

    pub fn estimate(&self) -> DataRate {
        DataRate::KilobitsPerSecFloat(self.estimate_kbps.unwrap_or_default())
    }

    fn Update(&mut self, capacity_sample: DataRate, alpha: f64) {
        let sample_kbps: f64 = capacity_sample.kbps_float();
        let estimate_kbps = if let Some(estimate_kbps) = self.estimate_kbps {
            (1.0 - alpha) * estimate_kbps + alpha * sample_kbps
        } else {
            sample_kbps
        };

        // Estimate the variance of the link capacity estimate and normalize the
        // variance with the link capacity estimate.
        let norm: f64 = estimate_kbps.max(1.0);
        let error_kbps: f64 = estimate_kbps - sample_kbps;
        self.deviation_kbps =
            (1.0 - alpha) * self.deviation_kbps + alpha * error_kbps * error_kbps / norm;
        // 0.4 ~= 14 kbit/s at 500 kbit/s
        // 2.5f ~= 35 kbit/s at 500 kbit/s
        self.deviation_kbps = self.deviation_kbps.clamp(0.4, 2.5);
        self.estimate_kbps = Some(estimate_kbps);
    }

    fn deviation_estimate_kbps(&self, estimate_kbps: f64) -> f64 {
        // Calculate the max bit rate std dev given the normalized
        // variance and the current throughput bitrate. The standard deviation will
        // only be used if self.estimate_kbps has a value.
        (self.deviation_kbps * estimate_kbps).sqrt()
    }
}
